src/encodings/Encodings.cpp

```diff
@@ -0,0 +1,389 @@
+		node = not(vars[inputLiterals[index].varID]);
+		history[std::pair<unsigned long, long>(inputLiterals[index].varID, curSum)] = SavedLit(0, inputLiterals[index].varID);
+	}
+	else {
+		node = varAlloc(auxVars, c);
+		if (!eq(low, true_lit)) {
+			formula = formula and (low or not(node));
+		}
+		if (eq(high, not(true_lit))) {
+			formula = formula and (not(vars[inputLiterals[index].varID]) or not(node));
+		}
+		else {
+			formula = formula and (high or not(vars[inputLiterals[index].varID]) or not(node));
+		}
+		history[std::pair<unsigned long, long>(inputLiterals[index].varID, curSum)] = SavedLit(1, auxVars.size()-1);
+	}
+	return node;
+}
+
+
+expr varAlloc(expr_vector& auxvars, context& c) {
+	static int nextVar = 0;
+	std::stringstream out;
+	out << "c_" << nextVar++;
+	auxvars.push_back(c.bool_const(out.str().c_str()));
+	return auxvars[auxvars.size() - 1];
+}
+
+std::string printBimanderVars(std::vector<std::vector<int>> vars) {
+	std::stringstream out;
+	for (auto& vec : vars) {
+		out << "[" << std::endl;
+		out << "\t";
+		for (auto& var : vec) {
+			out << var << " ";
+		}
+		out << std::endl;
+		out << "]" << std::endl;
+	}
+
+	return out.str();
+}
+
+
+std::string printNestedVars(std::vector<NestedVar> vars, int level) {
+	std::stringstream out;
+
+	int num = 1;
+	for (auto& var : vars)
+	{
+		if (var.varID != ULONG_MAX) 
+		{
+			out << var.varID << "-";
+		}
+		else
+		{
+			for (int i = 0; i < level && num>1; i++)
+			{
+				out << "\t";
+			}
+			out << " [ " << level << ":" << num++ << std::endl;
+			for (int i = 0; i < level+1; i++)
+			{
+				out << "\t";
+			}
+			out << printNestedVars(var.list, level + 1);
+			for (int i = 0; i < level; i++)
+			{
+				out << "\t";
+			}
+			out << " ] " << std::endl;
+		}
+
+	}
+	out << std::endl;
+	return out.str();
+}
+
+std::string printWeightedVars(std::vector<WeightedVar> wVars, expr_vector vars) {
+	std::stringstream out;
+	for (auto var : wVars) {
+		out << vars[var.varID].to_string() << " - " << var.weight << std::endl;
+	}
+	return out.str();
+}
+
+
+int findLongestPath(const CouplingMap cm, int nQubits){
+	connections.clear();
+	connections.resize(nQubits);
+	maxSum = -1;
+	for (auto edge: cm){
+		connections.at(edge.first).emplace_back(edge.second);
+	}
+	for (int q=0; q < nQubits; ++q){
+		d.clear();
+		d.resize(nQubits);
+		std::fill(d.begin(), d.end(), 0);
+		visited.clear();
+		visited.resize(nQubits);
+		std::fill(visited.begin(), visited.end(), false);
+		findLongestPath(q, 0);
+		auto it = std::max_element(d.begin(), d.end());
+		if ((*it)>maxSum)
+			maxSum = (*it);
+	}
+	return maxSum;
+}
+
+void findLongestPath(unsigned short node, int curSum){
+	if (visited.at(node))
+		return;
+	visited[node] = true;
+
+	if (d.at(node)<curSum)
+		d[node] = curSum;
+	if (connections.at(node).empty()){
+		visited[node]=false;
+		return;
+	}
+
+	for (auto child: connections.at(node)){
+		findLongestPath(child, curSum+1);
+	}
+
+	visited[node]=false;
```

Comment:
This should not be necessary. The Architecture class already instantiates a `distanceTable` at construction time which represents a Dijkstra tableau of the distances between all the nodes. It can be accessed via `architecture.getDistanceTable()` and the longest path is then given by the maximum over the 2D array.
Consequently the `Architecture` class should have a `getLongestPathRoutine` which does exactly this.